x/interchainqueries/keeper/grpc_query.go

```diff
@@ -31,19 +34,34 @@ func (k Keeper) RegisteredQueries(goCtx context.Context, req *types.QueryRegiste
 	return k.GetRegisteredQueries(ctx, req)
 }
 
-func (k Keeper) GetRegisteredQueries(ctx sdk.Context, _ *types.QueryRegisteredQueriesRequest) (*types.QueryRegisteredQueriesResponse, error) {
+func (k Keeper) GetRegisteredQueries(ctx sdk.Context, req *types.QueryRegisteredQueriesRequest) (*types.QueryRegisteredQueriesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.RegisteredQueryKey)
-	iterator := sdk.KVStorePrefixIterator(store, nil)
-	defer iterator.Close()
+	queries := make([]types.RegisteredQuery, 0, req.Pagination.Limit)
 
-	queries := make([]types.RegisteredQuery, 0)
-	for ; iterator.Valid(); iterator.Next() {
+	pageRes, err := querytypes.Paginate(store, req.Pagination, func(key, value []byte) error {
```

Comment:
i dont think the pagination works the way you expect. `onResult` callback for unmarshaling only. On the client side you will get responses with random amount of queries(which meets filter params) until the end of the iterator

Reply:
You are right. Please add a task to backlog with a full description of the issue and a possible solution (the one that we discussed on a call). We'll fix this next week.